### ISSUE TITLE
Support for curried functions

### ISSUE THREAD

**Issue by akkie** [jsdoc3/jsdoc#1286]
I've not found any documentation about how curried functions should be documented. Lets asume we have the following function:

```js
const push = browserHistory => route => browserHistory.push(route)
```

Is this a valid documentation?
```js
/**
 * @param {Object} browserHistory 
 * @param {string} route
 */
const push = browserHistory => route => browserHistory.push(route)
```

**Comment by kurtmilam** [comment]
I'm looking for the same thing. One option would be something like the following:
```javascript
/**
 * @param {Object} browserHistory
* @returns {Function} the function that will be called with the route
* @param {string} route
* @returns {???} whatever browserHistory.push(route) retunrs
*/
const push = browserHistory => route => browserHistory.push(route)
```
But I'd like to know whether there's an official way to do it.

**Comment by kurtmilam** [comment]
My IDE complains about the previous option, but seems OK with this one:
```javascript
/**
 * @param {Object} browserHistory - thing
 * @returns {Function} a function that accepts a single route parameter
 */
const push = browserHistory =>
  /**
   * @param {String} route - thing
   * @returns {Boolean} whatever browserHistory.push(route) returns
   */
  route => browserHistory.push(route)
```
Ugly, though.

**Comment by minexew** [comment]
Do you actually use curried functions in real-world code?

**Comment by kurtmilam** [comment]
@minexew Yes, I do actually use partial and curried functions in real-world code.

FYI, lodash, one of the most dependend-upon libraries in npm, offers `partial` and `curry` methods. Underscore offers a `partial` method. Native JavaScript offers a `partial` method in the form of `Function.prototype.apply`. JQuery even offers a method called `proxy` that can be used like `partial`, as well.

I prefer [Ramda](http://ramdajs.com/) and friends for my functional programming in JavaScript needs at the moment, but the new ES6 arrow function syntax also makes it quite easy to create your own curried functions (see previous posts in this discussion for an example).

**Comment by minexew** [comment]
All of those examples take in a "normal" function and currify it (or perform partial application). That's indeed often useful.

What I don't comprehend why anybody would want to make a function curried **by default**, in a way that doesn't allow supplying more than one argument per call.

**Comment by kurtmilam** [comment]
> What I don't comprehend why anybody would want to make a function curried by default, in a way that doesn't allow supplying more than one argument per call.

For the same reason they'd want to curry an existing function. It makes more sense when you get deeper into functional programming.

Furthermore, some libraries ([Ramda](http://ramdajs.com/) is one) offer a hybrid style of currying that lets you pass in one or several arguments at a time, and only calls the actual function after all of the arguments it needs have been supplied (for non-variadic functions with an arity <= 10).

So, the function in the original post is curried in a manner that requires one to call it twice, passing in one argument each time.

In Ramda-style, the function would be curried in a manner that would allow you to either pass in both arguments in one call or each argument in a separate call.

**Comment by renaudtertrais** [comment]
what about (just an idea):

```js
/**
 * @description filter xs if x[key] contains value
 * @param  {String}      key - key of xs items that should contains value
 * @return  {Function}
 * @param  {Any}          value - value that should be contain
 * @return  {Function}
 * @param  {Array}        xs - array ta must be filter
 * @return  {Array}        xs filtered
 */
const filterIfContains = key => value => xs => xs
  .filter(x => x[key].includes(value));

```

**Comment by Velinsky** [comment]
Does anyone actually have a solution for this instead of philosophical debates about the need of currying?

**Comment by jsatk** [comment]
> Does anyone actually have a solution for this instead of philosophical debates about the need of currying?

Agreed. Would love a solution for how to document this.

**Comment by karol-majewski** [comment]
This one makes Visual Studio Code provide proper IntelliSense:

```js
/**
 * @param {Object} browserHistory
 * @returns {function(string): boolean} a function accepting a single route parameter
*/
const push = browserHistory => route => browserHistory.push(route);
```

However, if you're using JSDoc annotations to leverage type checking, it won't do. To deal with TypeScript and its `Parameter 'route' implicitly has 'any' type.` message, you'd have to use inline types:

```js
const push = browserHistory => (/** @type {string} */ route) => browserHistory.push(route);
```

**Comment by dietergeerts** [comment]
Any news on this? Functional programming rocks in JS, makes everything more readable, understandable and testable, but the docs have to be ok too...

**Comment by angeloocana** [comment]
I'm looking for a solution to Ramda curry, I love it, but js docs is bad =/

Example:

```js
import { curry, filter, startsWith } from 'ramda';

/**
 * Get valid langKey in langs or return defaultLangKey
 * @func
 * @param {[String]} langs allowed lang keys ['en', 'fr', 'pt']
 * @param {String} defaultLangKey default browser language key
 * @returns {String} valid langKey
 */
const getValidLangKey = curry((langs, defaultLangKey, langKey) => {
  const currentLangKey = filter(l => startsWith(l, langKey), langs);
  return currentLangKey[0] || defaultLangKey;
});

export default getValidLangKey;
```

After adding **curry()** I lost all the types and descriptions.

I hope someone finds a good solution.

**Comment by asistapl** [comment]
@karol-majewski This solution provides best IntelliSense in VS Code:

![curriedfun](https://user-images.githubusercontent.com/18516855/32407040-e92f019a-c182-11e7-9d1f-9147d10e5363.png)

**Comment by karol-majewski** [comment]
@asistapl That's correct, support for JSDoc has gotten better over last year. I believe we can apply your solution to the original problem like this:

```js
// @ts-check

/**
 * @typedef {object} BrowserHistory
 * @property {(route: string) => void} push
 */

/**
 * @param {BrowserHistory} history
 * @returns {(route: string) => void}
 */
const push = history => route => history.push(route);
```

**Comment by eyalcohen4** [comment]
Would love to help with this one. arrow functions support will be great feature 😿

**Comment by lll000111** [comment]
Support in one IDE is one thing, I would be interested in support for the HTML output &mdash; in a way that makes the intent clear. For WebStorm I just wrote

```
/**
 * @param {MyTape} firstParam
 * @returns {function(function(SHA256Hash):boolean):function(SHA256Hash):Promise.<string>}
 */
```

which is a monster line (function that returns a function that returns a function). also, the ` @param` only is for the outermost parameter of the very first function. All others are declared in the `@return` tag, which gets very crowded.

The examples above look okay because there is only one level. It should work with 5 levels too (just to give a high(er) number).

**Comment by JesterXL** [comment]
For those of you still suffering, throwing 2 hacks into the ring.

The first is from the linked [Ramda issue](https://github.com/ramda/ramda/issues/2335). If you just put a type with a series of functions:

```javascript
/**
 * @type{function(cryptoModule): function(Function): function(Function) : function(string) : Promise<string>}
 */
const generateToken = curryN(4, (
        cryptoModule, 
        randomNumberInRangeFunction, 
        getHMACFunction,
        data
    ) =>
        getRandomNumber(randomNumberInRangeFunction)
        .then(num => getHMACFunction(cryptoModule, num, data))
)
```

Then you'll get a hint that works, but does not show the name of the parameters, just "arg0", "arg1", etc. At least it gets the return type correctly.

<img width="584" alt="screen shot 2018-02-24 at 12 06 03 pm" src="https://user-images.githubusercontent.com/424737/36632942-22a43460-195b-11e8-9a5f-c535c3d44597.png">

The second option is to just use a TypeScript definition file. This does not require the TypeScript compiler, nor to start using types. Downside here is your JSDocs are just for documentation; the intellisense is provided by the TypeScript definition file. Side good news is you can leave both and the `d.ts` file takes precedence, but doesn't affect your generated documentation.

Given the above function is in `index.js` and I create a `index.d.ts` file right next to it:

```typescript
export function generateToken(cryptoModule: object, randomNumberInRangeFunction: function, getHMACFunction: function, data: any): Promise<string>;
```

Now, I'll a code hint (in a file other than index.js):

<img width="584" alt="screen shot 2018-02-24 at 12 09 43 pm" src="https://user-images.githubusercontent.com/424737/36632986-a4aa674a-195b-11e8-87cd-a4bd786ba8a7.png">

**Comment by babakness** [comment]
Really look forward to seeing this happen!

**Comment by dietergeerts** [comment]
@minexew 
> What I don't comprehend why anybody would want to make a function curried by default, in a way that doesn't allow supplying more than one argument per call.

Because the rest of the code uses the function in different ways? like:

```
const doSomethingWith = _curry((configParam, dataParam) => {});
const doSomethingWithConfigA = doSomethingWith('lala');

doSomethingWithConfigA(5);
doSomethingWith('hey', 9);
```

The difficult part of this is adding JSDoc for `doSomethingWith`...

**Comment by nfantone** [comment]
What about point-free functions that don't have _explicit_ argument declaration?

```js
const { add } = require('ramda');

/**
*  Adds 10 to `x`.
*  @param {number} x Any number
*  @returns {number} The given `x` number plus 10.
*/
const sum10 = add(10);
```

☝️ This, unfortunately, doesn't work out of the box. Any ideas if this is even possible currently?

**Comment by dietergeerts** [comment]
@nfantone , it does work, but you have to add `@function` in the docblock

**Comment by nfantone** [comment]
@dietergeerts Worked like a charm. Thanks a million for pointing it out!

Still kinda awkward, if you ask me. I get the idea of wanting to infer types or match written docs with actual params to lint/give feedback on potentially bad tags - but _removing them entirely_ from the output seems a bit forced, honestly.

**Comment by JesterXL** [comment]
Y'all are wonderful, but I gave up trying to figure out how to get all these workarounds to work (insane nesting, jsdoc or babylon failing to recognize the `async` keyword, etc). For those who love types, I feel for you.

So I built a Hindley-Milner documentation library where you can use Markdown. Hopefully it helps y'all.

https://github.com/JesterXL/hm-doc

**Comment by joelnet** [comment]
@JesterXL The issue has only been open since October 2016. Be patient!

**Comment by lll000111** [comment]
@joelnet We get what we pay for...

**Comment by dietergeerts** [comment]
@JesterXL , like @joelnet says, be patient. Also keep in mind that there are features coming in JS regarding currying and partial application, which could change the implementation of this. (https://github.com/tc39/proposal-partial-application)

**Comment by phillt** [comment]
> Do you actually use curried functions in real-world code?

Yep, I use them create higher ordered components in React, and other various libraries for React use them including ReactRouter ReactRedux just to name a few.

**Comment by FernandoBasso** [comment]
> Do you actually use curried functions in real-world code?

Yes, I do.

**Comment by FernandoBasso** [comment]
Just as a note, with Haskell-like syntax for types, we could write something like this:

```
fn :: typeA -> typeB -> returnType
```

Alas, that is not how JavaScript parameter documentation is generally written.

It is also used in this on the most excellent [How to Design Programs](https://htdp.org/) book! It is nice because the notation is implicitly curried. Of course, it works better for certain languages than for others. Anyway, just wanted to mention it here. Someone may find it useful.

**Comment by SachinSaggezza** [comment]
Any update?

**Comment by BenjaminBrodwolf** [comment]
Still not possible to use JDoc on curried function?
How can I use JDoc with this example: `const pair = x => y => f => f(x)(y);`

**Comment by nfantone** [comment]
@BenjaminBrodwolf Well - nitpicking here, but that's not a curried function. That's a couple of function returning functions. So the "natural" way to jsdoc it, IMHO, would be:

```js
/**
 * Expects an argument and returns a function.
 * @param {*} x First argument.
 * @returns {Function} A function that expects an argument `y` and returns a new
 *  function that asks for a curried function `f` and invokes it with both `x` and `y`.
 */
const pair = x => y => f => f(x)(y);
```

Or you could document each returned function individually.

**Comment by kurtmilam** [comment]
@nfantone `const pair = x => y => f => f(x)(y)` is definitely a curried function.

> That's a couple of function returning functions.

That's the definition of true currying. Here, from [wikipedia](https://en.wikipedia.org/wiki/Currying) (emphasis mine):

> In mathematics and computer science, currying is the technique of translating the evaluation of a function that takes multiple arguments into evaluating a **sequence of functions, each with a single argument**. For example, a function that takes two arguments, one from X and one from Y, and produces outputs in Z, by currying is translated into **a function that takes a single argument from X and produces as outputs functions from Y to Z**.

You may be thinking of Ramda-style currying, which I've seen Ramda maintainers refer to as 'smart currying' to differentiate it from the actual currying @BenjaminBrodwolf 's code exemplifies.

Unfortunately, AFAIK, it's still not very convenient to document true currying via jsdoc.

**Comment by nfantone** [comment]
@kurtmilam Fair enough. Thanks for the clarification.

However, I stand by what I said. I was not actually thinking about Ramda or what brought up as "smart currying". Here, let me explain.

When we say that `f` is a "curried function" we are typically expressing "we can think of another function `g` of which `f` is _its curried form_". That's it. There's not actual "telling" if a function by itself is "curried", or "uncurried" or neither. A general statement such as that, makes little to no sense when context is removed.

So, in the context of JavaScript as a language, `const pair = x => y => f => f(x)(y)` **are some function returning functions** and, while I absolutely agree it could be improved, we do have a way of documenting that with jsdoc.

**Comment by lll000111** [comment]
@nfantone , @kurtmilam 

This was written for you guys: https://www.lesswrong.com/posts/7X2j8HAkWdmMoS8PE/disputing-definitions ("Disputing definitions")

The subject is how to document functions returning functions (returning functions). How you call this or that form is not even relevant in this particular context.

And similar discussions arise EVERY SINGLE TIME I see any discussion about functional programming things ALL those discussions devolve into people telling other people that they misunderstand some concept and that the _true_ definition is something else. The problem is, ! _definitions are 100% subjective_. There is no god handing them down. And it does not matter than someone somewhere wrote something down, even if they were first, or if they were famous. Human language mutates. People take something you say - and change it. The terms floating around functional programming are a prime example, where pretty much everybody forms their own ideas in their heads. I suggest to follow the advice in the linked blog post.

**Comment by kurtmilam** [comment]
@lll000111  Thanks for the link to the article. I have read and bookmarked it, and largely agree with it.

I'd like to clarify that I'm not a linguistic prescriptivist, and support an inclusive understanding of the word 'currying'. What I find problematic is someone else prescriptively declaring that a function that is curried by the classical definition is, in fact, not curried at all, especially in a discussion wherein better support for currying is being requested and discussed.

The statement is both prescriptivist and incorrect, and since this issue is about support for curried functions, I think it makes sense to be clear about whether a specific form of function fits the definition. Again, to be clear, I support an inclusive definition, and would like to see better support for classical currying as well as Ramda's 'smart' currying, which I also find useful and convenient.

As an aside, I'll point out that the article you linked to does include a qualifier about when it may make sense to argue about definitions:

> Whenever you feel tempted to say the words "by definition" in an argument that is not literally about pure mathematics

Currying is a mathematical term. It was coined by mathematicians and named after a mathematician ([Haskell Curry](https://en.wikipedia.org/wiki/Haskell_Curry)). It seems even the author of the piece you linked might suffer an argument about the true definition of currying :D

@nfantone It's unclear why you're still arguing the point. `const pair = x => y => f => f(x)(y)` is the curried version of `const pair = (x, y, f) => f(x)(y)` and `const pair = (x, y, f) => f(x)(y)` is the uncurried version of `const pair = x => y => f => f(x)(y)`, by (mathematical) definition. Neither the curried nor the uncurried version of the function is the 'true' version. They're just two different ways to express the function, one curried and one uncurried.

**Comment by nfantone** [comment]
Wow. This took a wild turn. Thanks both, @lll000111 and @kurtmilam for the time and fervent arguments.

@lll000111:

> The subject is how to document functions returning functions (returning functions). **How you call this or that form is not even relevant in this particular context**.

A 100% agreed. Let's stick to that.

@kurtmilam Let's just agree to disagree for now. I don't believe any of the (really good) resources you've shared strongly translates to your original, simple comment.

**Comment by nfantone** [comment]
On the topic, and since Ramda was mentioned before, an approach I usually take to document curried functions with jsdoc is to place the docs in the original/regular JS function and export the result of `curry`.

```js
const { curry, pathSatisfies } = require('ramda');

/**
 * Checks whether the given predicate is satisfied for the property value
 * at `propName` under `lambdaEvent.requestContext.authorizer`.
 *
 * @function
 * @see https://ramdajs.com/docs/#pathSatisfies
 * @param {Function} predicate The boolean returning function to apply to the
 *  value of the `propName` property.
 * @param {String} propName The name of the property to evaluate.
 * @returns {Boolean} The result of invoking `predicate` with the value of the
 *  property named `propName` on the `lambdaEvent.requestContext.authorizer` object.
 */
const authorizerPropSatisfies = curry(function authorizerPropSatisfies(predicate, propName) {
  return pathSatisfies(predicate, ['lambdaEvent', 'requestContext', 'authorizer', propName]);
});

module.exports = authorizerPropSatisfies;
```

**Comment by jirutka** [comment]
The name you’re looking for is high order functions, not currying. Just my two cents…

**Comment by MartinMuzatko** [comment]
It would be nice if this would work with VS Code
![image](https://user-images.githubusercontent.com/2950505/73278174-7e482a00-41eb-11ea-955f-a95ff79a27c7.png)


> On the topic, and since Ramda was mentioned before, an approach I usually take to document curried functions with jsdoc is to place the docs in the original/regular JS function and export the result of `curry`. Or, if I need the curried function inside the module I'm working (or happens to be a "private" function) just tag it with `@function`.
> 
> ```js
> const { curry, pathSatisfies } = require('ramda');
> 
> /**
>  * Checks whether the given predicate is satisfied for the property value
>  * at `propName` under `lambdaEvent.requestContext.authorizer`.
>  *
>  * @function
>  * @see https://ramdajs.com/docs/#pathSatisfies
>  * @param {Function} predicate The boolean returning function to apply to the
>  *  value of the `propName` property.
>  * @param {String} propName The name of the property to evaluate.
>  * @returns {Boolean} The result of invoking `predicate` with the value of the
>  *  property named `propName` on the `lambdaEvent.requestContext.authorizer` object.
>  */
> const authorizerPropSatisfies = curry(function authorizerPropSatisfies(predicate, propName) {
>   return pathSatisfies(predicate, ['lambdaEvent', 'requestContext', 'authorizer', propName]);
> });
> 
> module.exports = authorizerPropSatisfies;
> ```

**Comment by lll000111** [comment]
@MartinMuzatko 

> It would be nice if this would work with VS Code

Then the issue should be submitted to *that* team, no?

These are the issues for the documentation generator. First sentence from the README for this repo right here is:

> An API documentation generator for JavaScript.

If someone wants to work something with tool X I think it's better to ask the guys responsible for tool X. The author of this repo here can do nothing at all for VScode or whatever other IDE or tool.

**Comment by tonix-tuft** [comment]
> @karol-majewski This solution provides best IntelliSense in VS Code:
> 
> ![curriedfun](https://user-images.githubusercontent.com/18516855/32407040-e92f019a-c182-11e7-9d1f-9147d10e5363.png)

It seems that this syntax:

```js
/**
 * @param {BrowserHistory} history
 * @returns {(route: string) => void}
 */
const push = history => route => history.push(route);
```

Does not work well for JSDoc to Markdown generators like [jsdoc-to-markdown](https://github.com/jsdoc2md/jsdoc-to-markdown) and [markdox](https://github.com/cbou/markdox). These tools cannot parse these {(route: string) => void} expressions (it would be great they could).

Did anyone experience the same and come with a workaround when using such tools? How did you end up documenting your callbacks and thunks (higher-order function's function return value)?

**Comment by nfantone** [comment]
@MartinMuzatko 

> It would be nice if this would work with VS Code

It does work! Have you tried enabling `ts-check` for your sources?

![image](https://user-images.githubusercontent.com/476069/80871463-ad67af00-8ca4-11ea-8a0d-4fa749cb7472.png)

**Comment by tonix-tuft** [comment]
@nfantone How do you enable `ts-check` in VS Code?

**Comment by nfantone** [comment]
@tonix-tuft You can either add `// @ts-check` at the top of each `.js` file or set `"javascript.implicitProjectConfig.checkJs": true` on Settings to validate all sources.

**Comment by renatobenks** [comment]
there's still a discussion about if it worth doing this or it was decided that it should be supporting currying definitions?

**Comment by simonwjackson** [comment]
without adding an extra `type`, all parameters are assumed to be of type `any` 😭 

<img width="858" alt="Screen Shot 2020-07-01 at 1 15 16 PM" src="https://user-images.githubusercontent.com/189379/86277856-0bc5e680-bb9d-11ea-807f-82b870406562.png">

-----------------

This is the only solution that worked for me:

```js
const { curry, pathSatisfies } = require('ramda');

/**
 * Checks whether the given predicate is satisfied for the property value
 * at `propName` under `lambdaEvent.requestContext.authorizer`.
 *
 * @function
 * @see https://ramdajs.com/docs/#pathSatisfies
 * @param {Function} predicate The boolean returning function to apply to the
 *  value of the `propName` property.
 * @param {String} propName The name of the property to evaluate.
 * @returns {Boolean} The result of invoking `predicate` with the value of the
 *  property named `propName` on the `lambdaEvent.requestContext.authorizer` object.
 */

const authorizerPropSatisfies = curry(
  /** @type {(predicate: function, propName: string) => number} */
  (predicate, propName) => pathSatisfies(predicate, ['lambdaEvent', 'requestContext', 'authorizer', propName]),
);
```

<img width="932" alt="Screen Shot 2020-07-01 at 1 07 53 PM" src="https://user-images.githubusercontent.com/189379/86277225-f69c8800-bb9b-11ea-8e8b-ff0844bd25e8.png">

The tradeoff: you have to declare types twice, once inside jsdoc and once just above the function declaration (and inside `curry()`)

However, this doesn't feel completely unnatural as it resembles `function declarations` found in many FP languages

**Comment by nfantone** [comment]
@simonwjackson Ah, yes. You're absolutely right - I did not typed the arguments in my original examples for the inner function. From your suggestion, I noted that if you omit the type definitions for the top functions, vscode would infer them just fine nonetheless. 

```js
/**
 * Checks whether the given predicate is satisfied for the property value
 * at `propName` under `lambdaEvent.requestContext.authorizer`.
 *
 * @function
 * @see https://ramdajs.com/docs/#pathSatisfies
 * @param predicate The boolean returning function to apply to the
 *  value of the `propName` property.
 * @param propName The name of the property to evaluate.
 * @returns The result of invoking `predicate` with the value of the
 *  property named `propName` on the `lambdaEvent.requestContext.authorizer` object.
 */
const authorizerPropSatisfies = curry(
  /** @type {(predicate: (value: any) => boolean, propName: string) => (obj: Object) => boolean} */
  (predicate, propName) =>
    pathSatisfies(predicate, ['lambdaEvent', 'requestContext', 'authorizer', propName])
);
```

This way, there's no type duplication - at the probable expense of some reading clarity. I also improved some of the definitions so they are closer to their `@types/ramda` counterparts.

![image](https://user-images.githubusercontent.com/476069/86377395-33da4600-bc80-11ea-9a42-9071833cd382.png)

**Comment by simonwjackson** [comment]
@nfantone This is great! I wonder if the document generators will pick up the types as well.

**Comment by nfantone** [comment]
@simonwjackson Good question. Would have to try that - although my hunch says they won't. Let me know if you do!

**Comment by FernandoBasso** [comment]
> This one makes Visual Studio Code provide proper IntelliSense:
> 
> ```js
> /**
>  * @param {Object} browserHistory
>  * @returns {function(string): boolean} a function accepting a single route parameter
> */
> const push = browserHistory => route => browserHistory.push(route);
> ```
> 
> However, if you're using JSDoc annotations to leverage type checking, it won't do. To deal with TypeScript and its `Parameter 'route' implicitly has 'any' type.` message, you'd have to use inline types:
> 
> ```js
> const push = browserHistory => (/** @type {string} */ route) => browserHistory.push(route);
> ```

This solution worked for me with Vim, CoC and coc-tsserver. I'm very grateful! Thaks.

**Comment by chaorace** [comment]
Just wanted to share a `@callback` driven approach that I've been using. It preserves the type information pretty well and lets you keep the JSDoc blocks outside of your function definitions:
```
/**
 * @callback compareToRecordOrIdCb
 * @param {GlideRecord_proto|string} record A GlideRecord or sys ID string
 * @return {string}
 */

/** @type {function(Function): compareToRecordOrIdCb} */
const compareToRecordOrId = (operation) => (record) => {
	if (isGlideRecord(record)) {
		throwIfNoQueryResult(record);
		return operation(record.getUniqueValue());
	} else {
		return operation(record);
	}
};

/**
 * Provides a query string that will filter for exactly the given record
 *
 * @type {compareToRecordOrIdCb}
 */
export const exactlyThisRecord = compareToRecordOrId(hasSysId);
```

The main downside of this approach is that you will only see the callback type intellisense once you're inside of the actual parameter list, that info is missing when viewed in the finder list:

Parameter List (shows info):
![1599070129](https://user-images.githubusercontent.com/6462592/92020410-10865280-ed26-11ea-9c75-68938e456929.png)

Finder List (info missing):
![1599070111](https://user-images.githubusercontent.com/6462592/92020481-2bf15d80-ed26-11ea-8947-b474def1ea2f.png)

Keep in mind that `@callback` will only remember the description for `@param` tags, so you'll need to split your `@return` tag into two. You use one to document the return type inside of `@callback` and the other to document the return description inside of your main JSDoc block.

Here's a full and practical example with that fix applied:
```
/**
 * @callback compareToRecordOrIdCb
 * @param {GlideRecord_proto|string} record A GlideRecord or sys ID string
 * @return {string}
 */

/** @type {function(Function): compareToRecordOrIdCb} */
const compareToRecordOrId = (operation) => (record) => {
	if (isGlideRecord(record)) {
		throwIfNoQueryResult(record);
		return operation(record.getUniqueValue());
	} else {
		return operation(record);
	}
};

/**
 * Provides a query string that will filter for exactly the given record
 *
 * @return A query string to be used with query functions or GlideRecord.addEncodedQuery
 *
 * @throws {BadTableError} When record is a GlideRecord and is not configured for a valid table name
 * @throws {MissingRecordError} when record is a GlideRecord and is not indexed at any record
 *
 * @type {compareToRecordOrIdCb}
 */
export const exactlyThisRecord = compareToRecordOrId(hasSysId);

/**
 * Provides a query string that will filter out the given record
 *
 * @return A query string to be used with query functions or GlideRecord.addEncodedQuery
 *
 * @throws {BadTableError} When record is a GlideRecord and is not configured for a valid table name
 * @throws {MissingRecordError} when record is a GlideRecord and is not indexed at any record
 *
 * @type {compareToRecordOrIdCb}
 */
export const notThisRecord = compareToRecordOrId(notSysId);
```

Note how I'm also manually providing `@throws` in each main JSDoc block. That's because `@callback` only cares about the type signature. VS Code won't acknowledge most tags if you put them there (like `@throws` in this example)
